File: graduacao/orientacao_objetos/trabalho-i/Producoes/Livro.java
package Producoes;

/**
 * Producao que representa producoes do tipo Livro.
 * 
 * @author Gabriel S. Baptista
 *
 */
public class Livro extends Producao {
	private String editora;
	private String isbn;

	/**
	 * Construtor da Classe
	 * 
	 * @param titulo String representando o titulo.
	 * @param idioma String representando o idioma.
	 * @param cidade String representando a cidade.
	 * @param paginas Inteiro representando a quantidade de paginas.
	 * @param natureza String representando a natureza do documento.
	 * @param editora String representando a editora em que foi publicado.
	 * @param isbn String representando o isbn do documento.
	 */
	public Livro(String titulo, String idioma, String cidade, int paginas, String natureza, String editora,
			String isbn) {
		super(titulo, idioma, cidade, natureza, paginas);
		this.editora = editora;
		this.isbn = isbn;
	}
	
	/**
	 * Imprime o cabecalho da Classe livro.
	 */
	public static void imprimirCabecalho() {
		System.out.println("Natureza;Titulo;Idioma;Editora;Cidade;ISBN;Paginas");
	}
	
	/**
	 * Sobreescreve o metodo toString, ou seja, representa
	 * a maneira usual de se exibir uma producao do tipo livro.
	 */
	@Override
	public String toString() {
		String texto = this.getNatureza() + ";" + this.getTitulo() + ";" + this.getIdioma() + 
				";" + this.editora + ";" + this.getCidade() + ";" + this.isbn + ";";
		if(this.getPaginas() > 0) texto += this.getPaginas();
		return texto;
	}

	/**
	 * Sobreescreve o metodo equals.
	 */
	@Override
	public boolean equals(Object obj) {
		if(!(obj instanceof Livro)) return false;
		Livro referencia = (Livro) obj;
		if(this.compareTo(referencia) == 0) {
			return true;
		} else {
			return false;
		}
	}
	
	/**
	 * Sobreescreve o metodo compareTo, ou seja,  representa
	 * a maneira usual de se comparar producoes do tipo livro.
	 */
	@Override
	public int compareTo(Producao o) {
		Livro referencia = (Livro) o;
		// Verifica as caracteristicas
		int cmp = this.getNatureza().compareTo(referencia.getNatureza());
		if(cmp != 0) return cmp;
		cmp = this.getTitulo().compareTo(referencia.getTitulo());
		if(cmp != 0) return cmp;
		cmp = this.getIdioma().compareTo(referencia.getIdioma());
		if(cmp != 0) return cmp;
		cmp = this.editora.compareTo(referencia.editora);
		if(cmp != 0) return cmp;
		cmp = this.getCidade().compareTo(referencia.getCidade());
		if(cmp != 0) return cmp;
		cmp = this.isbn.compareTo(referencia.isbn);
		if(cmp != 0) return cmp;
		return this.getPaginas() - referencia.getPaginas();
	}
	
}

File: graduacao/orientacao_objetos/trabalho-ii/Livro.cpp
#include "Livro.hpp"

namespace producoes {
    /**
     *
     * @param titulo
     * @param idioma
     * @param cidade
     * @param natureza
     * @param paginas
     * @param tipo
     * @param edi
     * @param isbn_
     */
     Livro::Livro(const string& titulo, const string& idioma, const string& cidade,
      const string& natureza, const int& paginas, const int& tipo, const string& edi,
      const string& isbn_): Producao(titulo, idioma, cidade, natureza, paginas, tipo),
      editora(edi), isbn(isbn_) {}

    void Livro::print(ostream& out) const {
        out << natureza << ";" << titulo << ";" << idioma
            << ";" << editora << ";" << cidade << ";"
            << isbn << ";";
        //Verifica se paginas sera impresso
        if(paginas != 0) out << paginas;
    }

    bool Livro::operator<(const Producao& ref) {
        //Converte para Livro para poder acessar os dados
        Livro& o = (Livro&) ref;
        //Compara os dados individualmente
        int cmp = natureza.compare(o.natureza);
        if(cmp<0) return true; else if(cmp>0) return false;
        cmp = titulo.compare(o.titulo);
        if(cmp<0) return true; else if(cmp>0) return false;
        cmp = idioma.compare(o.idioma);
        if(cmp<0) return true; else if(cmp>0) return false;
        cmp = editora.compare(o.editora);
        if(cmp<0) return true; else if(cmp>0) return false;
        cmp = cidade.compare(o.cidade);
        if(cmp<0) return true; else if(cmp>0) return false;
        cmp = isbn.compare(o.isbn);
        if(cmp<0) return true; else if(cmp>0) return false;
        cmp = paginas - o.paginas;
        if(cmp<0) return true; else return false;
    }

}